docs/tap_hold.md

```diff
@@ -425,6 +425,167 @@ uint16_t get_quick_tap_term(uint16_t keycode, keyrecord_t *record) {
 If `QUICK_TAP_TERM` is set higher than `TAPPING_TERM`, it will default to `TAPPING_TERM`.
 :::
 
+## Chordal Hold
+
+Chordal Hold is intended to be used together with either Permissive Hold or Hold
+On Other Key Press. Chordal Hold is enabled by adding to your `config.h`:
+
+```c
+#define CHORDAL_HOLD
+```
+
+Chordal Hold implements, by default, an "opposite hands" rule. Suppose a
+tap-hold key is pressed and then, before the tapping term, another key is
+pressed. With Chordal Hold, the tap-hold key is settled as tapped if the two
+keys are on the same hand.
+
+Otherwise, if the keys are on opposite hands, Chordal Hold introduces no new
+behavior. Hold On Other Key Press or Permissive Hold may be used together with
+Chordal Hold to configure the behavior in the opposite hands case. With Hold On
+Other Key Press, an opposite hands chord is settled immediately as held. Or with
+Permissive Hold, an opposite hands chord is settled as held provided the other
+key is pressed and released (nested press) before releasing the tap-hold key.
+
+Chordal Hold may be useful to avoid accidental modifier activation with
+mod-taps, particularly in rolled keypresses when using home row mods.
+
+Notes:
+
+* Chordal Hold has no effect after the tapping term. 
+
+* Chordal Hold has no effect when the other key is also a tap-hold key. This is
```

Comment:
> Chordal Hold has no effect when the other key is also a tap-hold key.

Is this a limitation of the implementation?

Reply:
Thank you! Great comment. I was originally thinking multiple simultaneous tap-hold keys is out of scope of what Chordal Hold needs to consider, since in an input sequence like "`A`↓, `B`↓" where `A` and `B` are tap-hold keys, `A` and `B` can't be settled yet.

But rethinking it, these keys will eventually settle, of course, depending on subsequent events. I've revised so that Chordal Hold considers input sequences involving multiple tap-hold keys and I'm happy how this is working out.

Details:

1. Consider an input sequence "`A`↓, `B`↓, `C`↓, ..." of all presses of tap-hold keys. If held until the tapping-term, then regardless of handedness, these keys should be settled as held as usual. It is important to preserve this behavior for home row mods so that it is possible to chord multiple mods, e.g. Ctrl + Shift, in one hand.

2. Another case: consider if within the tapping term a key is released: "`A`↓, `B`↓, `C`↓, `C`↑". Provided either Permissive Hold or Hold on Other Key Press is enabled, the tap-hold keys `A`, `B` preceding `C` would then be settled as held. With Chordal Hold, perhaps depending on handedness some or all of the keys should be tapped instead.

   I've implemented a general heuristic (`waiting_buffer_find_chordal_hold_tap()` in the code) to decide this, based on evaluating `get_chordal_hold()` between successive pairs of keys. Some specific practical cases described for 3-key sequences, though the rule works for any number of keys:

   * If `A`, `B`, `C` are all on the same hand (`get_chordal_hold()` evaluates false between `AB` and `BC`), they are all considered tapped. This is effectively "typing streak" suppression of the hold function. This is cool!

   * If `A` and `B` are on one hand and `C` on the other, then both `A` and `B` are held.
   * If `A` is on one hand and `B` and `C` on the other, then just `A` is held.

5. Suppose an input sequence of tap-hold presses followed by a press of a regular key `Z`, like  "`A`↓, `B`↓, `C`↓, ..., `Z`↓". Then, all the same logic as in case 2 applies.

The implementation is more invasive than where this PR started, which makes me nervous. The "state machine" is not easy to reason about or modify, though I'm starting to get a feel for it.

I think you understand this area of the code a lot better than I do. Please let me know whether what I describe here needs further explanation, or seems like a flawed design and/or could be implemented better. I've made an effort to make it clean, been testing Chordal Hold in daily use (typing with it as we speak), and significantly beefed up the unit tests. But surely there's room for improvement. Anyway, I'm excited how this is progressing! 🔥

Reply:
Reading and changing the `action_tapping.c` file after a long hiatus is non-trivial (to say the least) and I'm trying to get up to speed again to evaluate this PR.

My user space implementation is similar to ZMK's "positional hold-tap" when the following are pressed sequentially within tapping term on the right side of a QWERTY layout:

1. Press RALT_T(KC_L)
2. Press RGUI_T(KC_K)
3. Continue holding both keys down

The host received the following:
```
KEY-DOWN - QMK: KC_L    Event key: l           Code: KeyL          KeyCode: 76
KEY-DOWN - QMK: KC_RGUI Event key: Meta        Code: MetaRight     KeyCode: 93
```

But Chordal implementation does not override the first key and will transmit both modifiers:
```
KEY-DOWN - QMK: KC_RALT Event key: Alt         Code: AltRight      KeyCode: 18
KEY-DOWN - QMK: KC_RGUI Event key: Meta        Code: MetaRight     KeyCode: 93
```
Will review the changes in more detail to figure this out.

I haven't considered the scenario of more than 2 simultaneous key presses, and it seems that specific use cases will have nuances. This PR might benefit from more end user tests to ensure robustness because support of this feature will land on the Discord server when merged.

Reply:
Thank you for the comparison! Yes, that is intended that both mods be held. I just added a unit test to verify this case. This behavior is useful so that one hand can hold Alt + GUI, or some other set of mods, while the other hand types a hotkey or uses the mouse.

OTOH, I can imagine there are other ways to handle those use cases. When using behavior as in your first output, what is a good solution for sending hotkeys like Ctrl+Shift+V? Do you find such multi-mod hotkeys are rare enough in practice that it's not a big deal? or maybe switch over to a Callum-style mod scheme for such things, or something else?

Reply:
Great point! You have convinced me. I use a layout with `STRD NEIO` on the home row (a [variant of Sturdy](https://getreuer.info/posts/keyboards/alt-layouts/index.html#magic-sturdy)), and yes, defaulting to same-hand tap behavior among those keys is compelling.

I've revised the PR to have this behavior and (after adding a few exemptions to my `get_chordal_hold()`) I like it a lot. Thank you for the suggestion!

Reply:
Thanks for making same-hand tap as default. Will provide feedback after testing this pull request as a daily driver.

Reply:
One problem I still have is if I add

```
bool get_chordal_hold(uint16_t tap_hold_keycode, keyrecord_t* tap_hold_record,
                      uint16_t other_keycode, keyrecord_t* other_record) {
    // Allow hold between any pair of mod-tap keys.
    if (IS_QK_MOD_TAP(tap_hold_keycode) && IS_QK_MOD_TAP(other_keycode)) {
        return true;
    }

    // Otherwise defer to the opposite hands rule.
    return get_chordal_hold_default(tap_hold_record, other_record);
}
```

to be able to trigger shift-ctrl-t (shift and ctrl one one hand, t on the other, with home row mods), that works fine but I now get on one hand rolls e.g. the alt-r I have on one of the home row keys applied, too.

Is there a way to fix that?

Reply:
@christoph-cullmann thank you for testing this out! I really appreciate it. And thank you for sharing your keymap.

You are right, as this PR currently works, there is a trade off about multi-mod hotkeys vs. protection from one-handed rolls. To be concrete, I'll consider specifically inputting the Shift-Ctrl-T hotkey on your keymap. This hotkey could be approached in the following ways:

* **Default behavior.** Supposing Chordal Hold is used with the default `get_chordal_hold()`, quickly pressing `RCTL_T(KC_A)` and then pressing `RSFT_T(KC_C)` (on the same hand) would incorrectly produce the first key as tapped rather than held.

  A workaround, which might be acceptable if Shift-Ctrl-T is an infrequent hotkey, is to press `RCTL_T(KC_A)`, wait out the tapping term, then press `RSFT_T(KC_C)`, and finally `LSFT_T(KC_T)`. This way, the first key gets settled as held as desired.

  Or in the other extreme if a multi-mod hotkey is especially frequent, creating a dedicated key (or combo, key override, tap dance, or leader key sequence) for it might be a good workaround. On that thought, I've been experimenting with a couple dedicated keys for Shift-Ctrl-left and Shift-Ctrl-right for text selection.

* **Exception between mod-taps.** Another possibility is to define `get_chordal_hold()` as you wrote to allow holds between pairs of mod-tap keys. In this case, quickly pressing `RCTL_T(KC_A)` and then `RSFT_T(KC_C)` settle as held as desired, but the downside now is that there is no protection from one-handed accidental mod triggers when rolling from one mod-tap to another mod-tap.

* **More selective exceptions.** A middle ground is to define `get_chordal_hold()` more selectively, in such a way that specific one-handed chords are allowed, rather than among all mod-taps. 

  ```c
  bool get_chordal_hold(uint16_t tap_hold_keycode, keyrecord_t* tap_hold_record,
                        uint16_t other_keycode, keyrecord_t* other_record) {
      // Exceptionally some one-handed chords.
      switch (tap_hold_keycode) {
          case RCTL_T(KC_A):  
              if (other_keycode == RSFT_T(KC_C)) {
                  // Allow hold in "RCTL_T(KC_A) down, RSFT_T(KC_C) down".
                  return true;
              }
              break;
          // Other exceptions...
      }
  
      // Otherwise defer to the opposite hands rule.
      return get_chordal_hold_default(tap_hold_record, other_record);
  }
  ```

  Then Shift-Ctrl-T can be typed quickly while still maintaining protection from one-handed accidental mod triggers otherwise.

> but I now get on one hand rolls e.g. the alt-r I have on one of the home row keys applied, too.

It looks like your keymap has defined the handedness of thumb keys to be `'*'`, if I followed that right:

```c
// special handle thumb keys
if (key.row == 3 || key.row == 7) return '*';
```

The `'*'` handedness means that holds are allowed in any chord involving that key, so that would remove the protection for rolls with the thumb R key. (I understand this is meant as one example, but just to note for Alt-R particularly.)

> Is there a way to fix that?

I'll give it some thought on what further can be done.

Reply:
Thanks a lot already for that detailed look into my issue.

You are right for the thumb key handling, that is intentional and I so far had no issues given I seem too slow to do proper rolls that involve the thumb :)

I read the above partial workarounds but I am not really sure they will help for multi-modifier shortcuts in general.

E.g. in Konsole a lot of the shortcuts are done that way to avoid clashes with the shell itself, therefore I don't see how one can special handle that in a good way. There shift-ctrl-t was just the one I picked as an example.

Ideally the implementation would delay all mod-tap decisions until a different handness key is pressed or the timeout sets in. But given my knowledge of the concrete implementation is zero, I assume that is a lot harder to do then just handle pairs.

I reverted atm back to use your great achordion user space lib :)

And I hope my comment doesn't lead to any blocking of this feature, beside this minor issue I have with multi modifier shortcuts vs. rolls all worked for me very well.

Reply:
I think there's a way! Looking further into this, I believe there is a solution to both provide same-hand roll protection as @filterpaper brought up and allow combining multiple same-hand modifiers.

It turns out that ZMK positional hold-tap ran into a similar limitation ([as urob described](https://github.com/urob/zmk-config?tab=readme-ov-file#timeless-homerow-mods), "it prevents combining multiple modifiers on the same hand"), and the solution suggested there is [an option](https://zmk.dev/docs/keymaps/behaviors/hold-tap#:~:text=By%20default%2C%20hold,key%20is%20tapped.) to delay until the next key *release* to decide how to settle: 

> By default, `hold-trigger-key-positions` are evaluated upon the first key press after the hold-tap. For homerow mods, this is not always ideal, because it prevents combining multiple modifiers unless they are included in `hold-trigger-key-positions`. To overwrite this behavior, one can set `hold-trigger-on-release`. If set to true, the evaluation of `hold-trigger-key-positions` gets delayed until key release. This allows combining multiple modifiers when the next key is held, while still deciding the hold-tap in favor of a tap when the next key is tapped.

Considered in the context of Chordal Hold, I think slightly better can be done, to reduce delay in some cases. Consider these two scenarios, where `RCTL_T(KC_A)` and `RSFT_T(KC_C)` are both on the right hand and `LSFT_T(KC_T)` is on the left hand:

```
                         TAPPING_TERM   
  +---------------------------|--------+
  | +---------------+         |        |
  | | RCTL_T(KC_A)  |         |        |
  | +---------------+         |        |
  |    +---------------+      |        |
  |    | RSFT_T(KC_C)  |      |        |
  |    +---------------+      |        |
  +---------------------------|--------+

                         TAPPING_TERM   
  +---------------------------|--------+
  | +-----------------------+ |        |
  | | RCTL_T(KC_A)          | |        |
  | +-----------------------+ |        |
  |    +--------------------+ |        |
  |    | RSFT_T(KC_C)       | |        |
  |    +--------------------+ |        |
  |        +--------------+   |        |
  |        | LSFT_T(KC_T) |   |        |
  |        +--------------+   |        |
  +---------------------------|--------+
```

Both scenarios start the same way, with pressing `RCTL_T(KC_A)` and then `RSFT_T(KC_C)`. Therefore, at the instant in time that `RSFT_T(KC_C)` is pressed, they cannot yet be distinguished. But it becomes clearer what ought to happen once a key is released or another key is pressed.

The logic I'm considering is something like the following:

1. In "`RCTL_T(KC_A)↓, RSFT_T(KC_C)↓, RCTL_T(KC_A)↑`" before the tapping term (top figure), `RCTL_T(KC_A)` is settled as tapped.
2. In "`RCTL_T(KC_A)↓, RSFT_T(KC_C)↓, RSFT_T(KC_C)↑`", both `RCTL_T(KC_A)` and `RSFT_T(KC_C)` are settled as tapped.
3. In "`RCTL_T(KC_A)↓, RSFT_T(KC_C)↓, KC_U↓`" (all keys on the same side), both `RCTL_T(KC_A)` and `RSFT_T(KC_C)` are settled as tapped.
4. In "`RCTL_T(KC_A)↓, RSFT_T(KC_C)↓, LSFT_T(KC_T)↓`", with the third key on the other side, we allow Permissive Hold or Hold On Other Keypress to decide how/when to settle the keys.
5. In "`RCTL_T(KC_A)↓, RSFT_T(KC_C)↓`" held until the tapping term, the keys are settled as held.

1&ndash;3 provide same-hand roll protection. 4&ndash;5 are for combining multiple same-hand modifiers.

I realize this is really getting into the nitty details, but let me know if there are any ideas or strong opinions about how to design these rules.

I think this above logic is something I can get implemented over the next week. I've gotten more comfortable with the action_tapping code, and this would resemble the "multi-tap" logic I had earlier on in this PR in commit https://github.com/qmk/qmk_firmware/pull/24560/commits/3fdbb079cbc28b2c4e8e85861c306f3669617dfa, but tightening it up for improved one-hand roll protection. Please let me know if you have suggestions.

Reply:
If you have something ready to test, I can give it an try :)
In any case, have nice Christmas days.

Reply:
@christoph-cullmann I've revised Chordal Hold to allow combining same-side mods, using with the rules in my previous message. I'm liking it!

I've been running it on my keyboard, for a few hours so far, and all seems good. This scheme allows combining same-side mods, yet it also has roll protection on streaks. For me, this feels like Achordion, but clearly better streak handling and improved responsiveness.

Thanks for the encouragement to make this change!

Reply:
Gave it some try on typing training via https://www.keybr.com/ - rolls seem really fine and no strange delays.

Reply:
Wonderful! That's great to hear. Thank you for the testing =)

Reply:
Tried this diff today. This actually is much smoother that Achordion mode which was adding annoying delays on the home row mod. Cannot wait for it to land upstream!

Reply:
Did use it now the last days full time and so far found no issues, at least for my use case with the home row mods this state fully replaces Achordion, thanks a lot. Hope it can be merged without any issues.

Reply:
@alxzh @christoph-cullmann thank you both for the feedback! I agree, qualitatively this resembles and improves upon Achordion.